### PR TITLE
Custom Pages add initial unittests

### PR DESCRIPTION
**Pull request by kingthorin**
Add UnitTests for new `AbstractPlugin` functionality, `CustomPageMatcherLocation`, `CustomPage.Type`, `DefaultCustomPage`
encode/decode functionality.

Signed-off-by: kingthorin <kingthorin@users.noreply.github.com>

### CODE REVIEW

File: zap/src/main/java/org/zaproxy/zap/extension/custompages/CustomPage.java [hunk 1]
```diff
@@ -149,6 +150,10 @@ public String getName() {
             return name;
         }
 
+        public String getNameKey() {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/custompages/CustomPage.java:153]: Could be just package accessible, just for tests.

File: zap/src/main/java/org/zaproxy/zap/extension/custompages/CustomPageMatcherLocation.java [hunk 2]
```diff
@@ -50,14 +51,16 @@ public int getId() {
...
      *
-     * <p>
-     *
      * @return the name of the custom page page matcher location
      */
     public String getName() {
         return name;
     }
 
+    public String getNameKey() {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/custompages/CustomPageMatcherLocation.java:60]: Same here.

File: zap/src/main/java/org/zaproxy/zap/extension/custompages/DefaultCustomPage.java [hunk 3]
```diff
@@ -267,6 +268,7 @@ protected static DefaultCustomPage decode(int contextId, String encodedString) {
                             Type.getCustomPageTypeWithId(Integer.parseInt(pieces[3])), // Type
                             Boolean.parseBoolean(pieces[4])); // Enabled
         } catch (Exception ex) {
+            System.out.println(ex.getMessage());
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/custompages/DefaultCustomPage.java:271]: Remove.

File: zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java [hunk 4]
```diff
@@ -1220,6 +1243,101 @@ public void shouldSaveToConfig() {
                 is(equalTo(Plugin.AttackStrength.INSANE.name())));
     }
 
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void shouldCheckPage200WithParent(boolean expectedResult) {
```
Comment (thc202) [inline zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java:1248]: The parameter could be removed (and this be a normal test), not used.

File: zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java [hunk 5]
```diff
@@ -1220,6 +1243,101 @@ public void shouldSaveToConfig() {
...
+        assertThat(result, is(equalTo(expectedResult)));
+        verify(parent).isCustomPage(message, type);
+        verify(analyser).isFileExist(message);
+    }
+
+    @Test
+    public void shouldCheckPage404WithParent() {
+        // Given
+        CustomPage.Type type = CustomPage.Type.NOTFOUND_404;
+        ;
```
Comment (thc202) [inline zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java:1283]: Remove.

File: zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java [hunk 6]
```diff
@@ -1220,6 +1243,101 @@ public void shouldSaveToConfig() {
...
+        // When
+        boolean result = plugin.isPage404(message);
+        // Then
+        assertThat(result, is(equalTo(true)));
+        verify(parent).isCustomPage(message, type);
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void shouldCheckPage404WithParentWithFallbacktoAnalyser(boolean expectedResult) {
```
Comment (thc202) [inline zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java:1295]: `AndFallback`?

File: zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java [hunk 7]
```diff
@@ -1220,6 +1243,101 @@ public void shouldSaveToConfig() {
...
+        // Given
+        CustomPage.Type type = CustomPage.Type.NOTFOUND_404;
+        ;
```
Comment (thc202) [inline zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java:1298]: Remove.

File: zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java [hunk 8]
```diff
@@ -1220,6 +1243,101 @@ public void shouldSaveToConfig() {
...
+        verify(parent).isCustomPage(message, type);
+        verify(analyser).isFileExist(message);
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    public void shouldCheckPage500WithParent(boolean expectedResult) {
+        // Given
+        CustomPage.Type type = CustomPage.Type.ERROR_500;
+        ;
```
Comment (thc202) [inline zap/src/test/java/org/parosproxy/paros/core/scanner/AbstractPluginUnitTest.java:1317]: Remove.

File: zap/src/test/java/org/zaproxy/zap/extension/custompages/DefaultCustomPageUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.parosproxy.paros.Constant;
+import org.zaproxy.zap.utils.I18N;
+
+class DefaultCustomPageUnitTest {
+
+    private static final String TEST_MATCH_PATTERN = ".*Something went wrong.*";
+    private static final String BASE64_TEST_PATTERN =
+            new String(Base64.getEncoder().encode(TEST_MATCH_PATTERN.getBytes()));
```
Comment (thc202) [inline zap/src/test/java/org/zaproxy/zap/extension/custompages/DefaultCustomPageUnitTest.java:35]: `getBytes(StandardCharsets.UTF_8)), StandardCharsets.US_ASCII`

**Comment by kingthorin** [comment]
Seems GitHub is having some issues right now. If you look at https://github.com/zaproxy/zaproxy/compare/develop...kingthorin:cp-unittests you can see that the branch was updated b9224eb, however, the PR is still showing 408e005 (currently).

- https://www.githubstatus.com/incidents/28js6274ybpb?utm_ts=1602280438
- https://twitter.com/githubstatus

**Comment by kingthorin** [comment]
Seems sorted out now.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]

**Pull request merged**